### PR TITLE
DAT-3579 add wikia creation event, event producer refactor

### PR DESCRIPTION
**Pull request by idradm**
https://wikia-inc.atlassian.net/browse/DAT-3579

### CODE REVIEW

File: extensions/wikia/CreateNewWiki/CreateWiki.php [hunk 1]
```diff
@@ -422,6 +422,8 @@ public function create() {
 
 		wfDebugLog( "createwiki", __METHOD__ . ": Local maintenance task added as {$task_id}\n", true );
 
+		wfRunHooks( "afterWikiCreated", [ $this->mNewWiki->city_id, $this->sDbStarter ] );
```
Comment (sebastianmarzjan) [inline extensions/wikia/CreateNewWiki/CreateWiki.php]: Let's uppercase the first letter of hook name for consistency

File: extensions/wikia/IndexingPipeline/IndexingPipeline.setup.php [hunk 2]
```diff
@@ -1,23 +1,26 @@
...
-$wgHooks['ArticleDeleteComplete'][] = 'Wikia\IndexingPipeline\PipelineEventProducer::onArticleDeleteComplete';
-$wgHooks['ArticleUndelete'][] = 'Wikia\IndexingPipeline\PipelineEventProducer::onArticleUndelete';
-$wgHooks['TitleMoveComplete'][] = 'Wikia\IndexingPipeline\PipelineEventProducer::onTitleMoveComplete';
-$wgHooks['UserTemplateClassification::TemplateClassified'][] = 'Wikia\IndexingPipeline\PipelineEventProducer::onTemplateClassified';
+$wgHooks[ 'NewRevisionFromEditComplete' ][] = 'Wikia\IndexingPipeline\PipelineEventProducer::onNewRevisionFromEditComplete';
+$wgHooks[ 'ArticleDeleteComplete' ][] = 'Wikia\IndexingPipeline\PipelineEventProducer::onArticleDeleteComplete';
+$wgHooks[ 'ArticleUndelete' ][] = 'Wikia\IndexingPipeline\PipelineEventProducer::onArticleUndelete';
+$wgHooks[ 'TitleMoveComplete' ][] = 'Wikia\IndexingPipeline\PipelineEventProducer::onTitleMoveComplete';
+$wgHooks[ 'UserTemplateClassification::TemplateClassified' ][] = 'Wikia\IndexingPipeline\PipelineEventProducer::onTemplateClassified';
+$wgHooks[ 'afterWikiCreated' ][] = 'Wikia\IndexingPipeline\PipelineEventProducer::onAfterWikiCreated';
```
Comment (sebastianmarzjan) [inline extensions/wikia/IndexingPipeline/IndexingPipeline.setup.php]: See https://github.com/Wikia/app/pull/9412/files#diff-0592b2de34e1a03712c20967015312d4R425

File: extensions/wikia/IndexingPipeline/PipelineEventProducer.class.php [hunk 3]
```diff
@@ -58,16 +60,22 @@ public static function sendFlaggedSyntax( $action, $pageId, $revisionId, $ns = s
...
-		$ns = self::preparePageNamespaceName( $article->getTitle() );
-		$action = $rev->getPrevious() === null ? self::ACTION_CREATE : self::ACTION_UPDATE;
+		$action = $rev->getPrevious() === null ? PipelineRoutingBuilder::ACTION_CREATE
+			: PipelineRoutingBuilder::ACTION_UPDATE;
 		$pageId = $article->getId();
 		$revisionId = $rev->getId();
 
 		self::send( 'onNewRevisionFromEditComplete', $pageId, $revisionId );
-		self::sendFlaggedSyntax( $action, $pageId, $revisionId, $ns );
+		self::sendFlaggedSyntax( $action, $pageId, $revisionId, $article->getTitle()->getNamespace() );
```
Comment (sebastianmarzjan) [inline extensions/wikia/IndexingPipeline/PipelineEventProducer.class.php:78]: Is it possible to get non-article or non-title here? ( I see it wasn't checked for previously, so the working assumption is: no )

File: extensions/wikia/IndexingPipeline/PipelineRoutingBuilder.class.php [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+	/**
+	 * Adds action key to routing
+	 *
+	 * @param string $action
+	 *
+	 * @return PipelineRoutingBuilder $this
+	 */
+	public function addAction( $action ) {
+		if ( in_array( $action, [ self::ACTION_CREATE, self::ACTION_DELETE, self::ACTION_UPDATE ] ) ) {
```
Comment (sebastianmarzjan) [inline extensions/wikia/IndexingPipeline/PipelineRoutingBuilder.class.php]: I think t'd make sense to extract the condition and do if ($this->isSupportedAction( $action ) ). Maybe also throw on invalid one?

File: extensions/wikia/IndexingPipeline/PipelineRoutingBuilder.class.php [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+	 * @return PipelineRoutingBuilder $this
+	 */
+	public function addName( $name ) {
+		$this->name = $name;
+
+		return $this;
+	}
+
+	public function addType( $type ) {
+		if ( in_array( $type, [ self::TYPE_ARTICLE, self::TYPE_WIKIA ] ) ) {
```
Comment (sebastianmarzjan) [inline extensions/wikia/IndexingPipeline/PipelineRoutingBuilder.class.php]: I think t'd make sense to extract the condition and do if ($this->isSupportedType( $type ) ).
Maybe also throw on invalid one?

File: extensions/wikia/IndexingPipeline/PipelineRoutingBuilder.class.php (hunk 4 above)
Comment (idradm) [inline extensions/wikia/IndexingPipeline/PipelineRoutingBuilder.class.php]: Dedicated one? or runtime is enough?

Comment (sebastianmarzjan) [inline extensions/wikia/IndexingPipeline/PipelineRoutingBuilder.class.php]: Builtin runtime / invalid argument exception will be fine

**Comment by dianafa** [comment]
looks nice

File: extensions/wikia/IndexingPipeline/tests/PipelineRoutingBuilderTest.php [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+	public function testAllArgs() {
+		$expected = "test._action:create._namespace:content";
+		$this->assertEquals( $expected, PipelineRoutingBuilder::create()
+			->addName( "test" )
+			->addAction( PipelineRoutingBuilder::ACTION_CREATE )
+			->addNamespace( 0 )
+			->build() );
+	}
+
+	public function testAllArgsInRandomOrder() {
```
Comment (sebastianmarzjan) [inline extensions/wikia/IndexingPipeline/tests/PipelineRoutingBuilderTest.php:29]: :heart:

**Comment by sebastianmarzjan** [comment]
LGTM

**Pull request merged**